### PR TITLE
Fix some Unicode issues with Python 2/3 by using six. Ignore Vim undo files

### PR DESCRIPTION
**Pull request by prologic**

### CODE REVIEW

**Comment by TomCrypto** [comment]
Thanks a bunch, it works much better with Python 2 now! It's still not working properly with input files that contain non-ascii characters however, like [this random github readme](https://raw.githubusercontent.com/unitedstates/congress-legislators/master/README.md), it would throw an unicode decoding exception inside python-markdown. I was able to get it to work using:

``` python
f = fileinput.input(files=args.file, mode='rb')
text = ''.join([line.decode('utf-8') for line in f])
stream = markdown.markdown(text, exts)
f.close()
```

i.e. bypassing fileinput's encoding completely and just decoding the whole damn file into an utf-8 string myself. Miraculously this works in both Py2 and Py3, so I'm pretty okay with it. The other thing was the literal unicode symbols in `themes.py` were getting garbled when printed to the terminal, had to convert them to e.g. `\u2022` and such, and making all the strings in there unicode as well. After that it seemed to work properly all the time.

What do you think? It looks pretty clean other than the final print with the python version check which is a bit icky, but other than that...

**Comment by prologic** [comment]
Basically if you have a read of my gist [Python 2/3 Unicode](https://gist.github.com/prologic/24cafc7315596b8683ca) you just need to use `u()` everywhere you expose a public API. That is everywhere you use a string; use `u()`.

Read files as bytes and decode appropriately. i.e: `open("foo.txt", "rb").read().decode("utf-8")`

`print()` will do the right thing with Unicodeo bjects; so don't worry too much about encoding into bytes (_unless you're dealing with sockets_).

`u("foo {1}").format(u("bar"))` also works as a pattern to.

Basically don't use `"foo"` anywhere in your code :) In Python 2 these are bytes/ascii; in Python 3 they are strings/unicode!

**Comment by TomCrypto** [comment]
That makes sense, I'll make sure to use unicode strings everywhere with `u()`. However I observe that `print` does _not_ do the right thing when stdout is piped, which is the last thorny issue that I think we are having. For instance, with Python 2 if I put a `print(type(renderer.render(markup)))` before printing out the final rendered markdown, to make sure it's actually unicode, I get:

```
$ lanius README.md
<type 'unicode'>
(rendered markdown, displays properly)
```

However if I pipe it into less for instance:

```
$ lanius README.md | less -R
Traceback (most recent call last):
  File "/usr/local/bin/lanius", line 9, in <module>
    load_entry_point('lanius==0.1.2', 'console_scripts', 'lanius')()
  File "build/bdist.linux-x86_64/egg/lanius/runner.py", line 98, in main
<type 'unicode'>
  File "build/bdist.linux-x86_64/egg/lanius/runner.py", line 88, in run
UnicodeEncodeError: 'ascii' codec can't encode characters in position 2-28: ordinal not in range(128)
```

In Python 3 this works fine, in Python 2 I need to `encode('utf-8')` my string before passing it to `print()` for it to display properly. But of course I can't `encode()` in Python 3, otherwise I end up printing a literal byte sequence. That's why I have this if/else case when printing the final output, if that makes sense.

**Comment by prologic** [comment]
I just pushed a fix this.

Basically you are using the Python 2 `print` statement; not the new function that was introduced in Python 2.7 something. It is a function in Python 3 :)

So:

``` #!python
from __future__ import print_function
```

At the top of your module(s) wher eyou want to use the `print(0` function.

However if you just use `sys.stdout.write()` you'll be fine in both cases I believe :) Or use the `print()` function!

**Comment by TomCrypto** [comment]
Hm, no, still doesn't work on my side when piping the output, either with the print_function or sys.stdout.write. How to do it without different code in python2 and python3 is eluding me. Will look into it further this afternoon.

**Comment by prologic** [comment]
Works for me but some of your unicide char don't display very welll :)

James Mills / prologic

E: prologic@shortcircuit.net.au
W: prologic.shortcircuit.net.au

**Comment by TomCrypto** [comment]
Thanks! Merged, with a couple commits to fix the unicode chars displaying funnily as well as the issue I was having. How does this look now?

Note the theme is not definitive, it was quickly written up to get something nice on the screen... basically the theme.py is kind of a mess right now till I figure out a flexible way to expose "hooks" into the renderer! :)

**Comment by prologic** [comment]
All good :+1: -- [Tenets of the UNIX Philosophy #3](http://www.ru.j-npcs.org/usoft/WWW/LJ/Articles/unixtenets.html) "Build a prototype as soon as possible"
